### ISSUE TITLE
[HOLD for payment 2023-10-23] [HOLD for payment 2023-10-23] [HOLD for payment 2023-10-20] [HOLD for payment 2023-10-20] [$1000] Web - When input exceeds 500k characters, the web freezes.

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#25804]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to Create a New Chat, New Group, or New Task.
2. Input exceeds 500k characters into the search bar or title.

## Expected Result:
The website does not crash when entering long text.

## Actual Result:
 The website is frozen or crash

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** v1.3.56-21

**Reproducible in staging?:**  Y

**Reproducible in production?:** Y

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/115492554/3a6afc9b-efd0-47a6-8960-15c12845a5b6

https://github.com/Expensify/App/assets/115492554/17a0a4d2-e946-4b8e-9e61-f3bcdc8f288c

https://github.com/Expensify/App/assets/115492554/60c3fe56-7374-4fc7-95a5-b37059ab4319

**Expensify/Expensify Issue URL:**

**Issue reported by:** @LeThiThuThuy

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1691988614374279

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ffb0be0940793404</li>
        <li>Upwork Job ID: 1696326036875673600</li>
        <li>Last Price Increase: 2023-09-19</li>
    </ul>
</details>

**Comment by ahmedGaber93** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When input exceeds 500k characters, the web freezes.
### What is the root cause of that problem?
There is no `maxLength` prop in OptionsSelector which make TextInput accept infinity length of text
### What changes do you think we should make in order to solve the problem?
We need to add `maxLength` prop to OptionsSelector here
https://github.com/Expensify/App/blob/bfe5c21588042958926a1bcb22239635e8b379fb/src/pages/NewChatPage.js#L192

We can pass it as CONST with 1000 or what we agree on.


### What alternative solutions did you explore? (Optional)
N/A

**Comment by jeet-dhandha** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
- Issue is that on pasting longer text in composer, it slows down the browser. As processing longer text takes time, low end devices are affected more.

### What is the root cause of that problem?
- The root cause is that the text is processed in one go. So, the browser is busy processing the text and can't do anything else.

### What changes do you think we should make in order to solve the problem?
- We can add a limit on paste event. So, if the text is longer than the limit we can show `Paste Limit Exceeded` error.
- Another option is to clip the text to the limit and show `Paste Limit Exceeded` error.

- Finally remove the `Paste Limit Exceeded` error after `1.5s`.

<img width="1440" alt="Screenshot 2023-08-24 at 9 39 10 AM" src="https://github.com/Expensify/App/assets/78416198/0efbcaa0-e4dd-4b50-930f-93c574c5933e">


### What alternative solutions did you explore? (Optional)
- N/A

**Comment by johncschuster** [comment]
Not overdue. The issue has just been triaged to `External`.

**Comment by meldevigolsen** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Inputting too many characters in the input for contact search causes the page to crash and require a refresh of the page.



### What is the root cause of that problem?

RegExp throws this uncaught error:

![image](https://github.com/Expensify/App/assets/31360158/dd4d8d45-263f-43dc-a908-e29fe1901496)

![image](https://github.com/Expensify/App/assets/31360158/92ad7424-db4a-4d3c-b72e-aa31757bec40)

Further debugging shows that the error is thrown from this line:

https://github.com/Expensify/App/blob/cdec13da85ea64b735ba8717860feb4e462b9323/src/libs/OptionsListUtils.js#L561


### What changes do you think we should make in order to solve the problem?

Add a string length check in `src/libs/OptionsListUtils.js` and return a `common.characterLimit` message if string length exceeds 5000 characters.

https://github.com/Expensify/App/blob/cdec13da85ea64b735ba8717860feb4e462b9323/src/languages/en.js#L107

This will inform the user of the limitations and prevent this from crashing the page. This will also take advantage of preexisting and future translations.

### What alternative solutions did you explore? (Optional)
(Blank)

**Comment by meldevigolsen** [comment]
Contributor details
Your Expensify account email: meldevigolsen@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01812d3213d2b1075f

**Comment by jeet-dhandha** [comment]
I know my proposal isn't directly solving the issue. As main issue is the freezing of web which is mostly due to pre-processing of emojis.

**Comment by jeet-dhandha** [comment]
My bad 😂 i had worked so much on Composer lately, thus it might be stuck in mind as a Composer's issue at that time.

**Comment by dannymcclain** [comment]
I agree with @trjExpensify – I think it would be smart and good to add reasonable character limits to All The Things™ and then include a standard error message when the user goes over that limit. Something _kinda like_ this but not necessarily _exactly_ this. I think putting in the limits and error messages will just make things more predictable in general, and helpful when things go wrong.

<img width="676" alt="image" src="https://github.com/Expensify/App/assets/7516624/0cb24617-c838-48cd-b314-f94dbcaf5755">

**Comment by shawnborton** [comment]
Nice, that makes sense to me.

**Comment by joekaufmanexpensify** [comment]
Agreed. I hadn't had the chance to bring this to Slack yet, but this was exactly what I was thinking. A few Q's for discussion: 

- Would we show the counter when there's no error? (Personally I think no). 
- Would we still add some upper bound of what you can put into the field before the cursor cuts you off? (I feel like it could make sense to. Otherwise, I'm sure we'll see other bug reports where the app breaks because of pasting something huge like ~500k characters into a field). 
- When would we remove the error? (Looking at other errors in NewDot, we seem to remove them right when they're resolved, rather than on blur. So I'm thinking we remove the error as soon as you hit or drop below the character limit). 

Curious to hear other's thoughts on these though!

**Comment by dannymcclain** [comment]
@joekaufmanexpensify Here are my personal thoughts on your bullets, definitely curious what others will have to say too:

- I would say no, don't show the counter unless there's an error. I think for the _most_ part people aren't going to encounter this at all, so keep it out of the way until someone needs to see it
- Still having an upper bound that stops your input does makes sense—I think I would defer to someone more technical to weigh in on that
- I agree with you, to me it makes sense to remove the error as soon as you hit/drop below the limit. I think that typically we do validate on blur but I feel like in this case it makes sense to validate on change once that error is thrown. This is one of those cases where immediate feedback makes sense.

**Comment by shawnborton** [comment]
> I would say no, don't show the counter unless there's an error. I think for the most part people aren't going to encounter this at all, so keep it out of the way until someone needs to see it

> I agree with you, to me it makes sense to remove the error as soon as you hit/drop below the limit. I think that typically we do validate on blur but I feel like in this case it makes sense to validate on change once that error is thrown. This is one of those cases where immediate feedback makes sense.

Definitely agree with these!

**Comment by joekaufmanexpensify** [comment]
I was thinking of pasting here specifically. I generally agree with not truncating pasted content. But at a certain (very high limit), I feel like there is an argument to do that to prevent performance issues. 

I don't think it's realistic that any legitimate user would be trying paste something like 500k characters into a field. And even if we solve the performance concerns with 500k characters, we could just see the same issue with a higher amount of characters (if there's no upper bound on what you can paste in). 

I don't feel too strongly about that though, we could always do nothing for that now and see if it comes up again later.

**Comment by trjExpensify** [comment]
>I don't think it's realistic that any legitimate user would be trying paste something like 500k characters into a field. And even if we solve the performance concerns with 500k characters, we could just see the same issue with a higher amount of characters (if there's no upper bound on what you can paste in).

Do we know what the upper bound is that causes a crash? If it's like 100k, we could potentially set a upper bound that just auto-truncates instead of crashes the app, and the character limit exceeded error message will throw with the counter for feedback.

What I want to try and avoid really is auto-truncating "on the limit" of say 50 characters. So you paste 70 characters and it cuts off 20 of those. So to "rework" what you want to paste you end up having to do that elsewhere to get below the limit and then re-paste. I think it's a better experience if you can "edit in place" with the character count guidance as you do.

**Comment by joekaufmanexpensify** [comment]
For sure. That makes a lot of sense to me. I'm not sure what the upper bound is for crash right now, but I expect it is something large like 100k. Once we know, I like auto-truncating (instead of crashing the app) in that one specific instance. And then otherwise allowing you to exceed the limit editing in place. 

And as long as the upper bound is really large (like 100K characters), I doubt we'll run into many/any legit scenarios where a customer is pasting more than this into the field.

**Comment by joekaufmanexpensify** [comment]
I was thinking we'd add the new field limit behavior here. So add [the counter](https://github.com/Expensify/App/issues/25804#issuecomment-1705685194) if you go over the limit, and then an upper bound limit where we block adding any additional characters to the field.

**Comment by joekaufmanexpensify** [comment]
Yep, it's included in the hyperlink in my above message!

**Comment by ahmedGaber93** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/25804#issuecomment-1690722686)

**Comment by ahmedGaber93** [comment]
I think displaying the error with maxLength number without counter will be simpler UI and in the code like that

![Screenshot 2023-09-14 at 11 01 52 PM](https://github.com/Expensify/App/assets/41129870/88566567-a2c4-480c-9489-bf1034233e91)

@joekaufmanexpensify  @allroundexperts

**Comment by joekaufmanexpensify** [comment]
>For the scenario where the user has entered between 50 and 70 characters, do we want to show the counter message as an error (in red colour)? The screenshot https://github.com/Expensify/App/issues/25804#issuecomment-1705685194 is giving me the impression that we want to do that.

Yep!

>When the user enters more than 70 characters, do we still show them a counter? Or should the message be different?

I don't think the user would be able to type more than 70 characters. If they did, I think it'd be fine to keep the same error, and just block them from entering more characters. Then, if they pasted more than 70 characters, we'd truncate, and show the counter error capped at 70.

**Comment by joekaufmanexpensify** [comment]
>I think displaying the error with maxLength number without counter will be simpler UI and in the code editing, and immediately prevent the user to type more text like that. Although without counter, we will disable OptionsSelector from filter the data with the exceed text because we immediately prevent the user from typing after exceeded.
What do you think about that? 

I think it's fine to go no counter when the user hits the max length. If they then reduce the input by any amount of characters, we'd immediately show them the counter then (rather than changing the error on blur)?

**Comment by ahmedGaber93** [comment]
@allroundexperts We can do this [behavior](https://github.com/Expensify/App/issues/25804#issuecomment-1726788809) by this edit.
<details>
<summary>code changes</summary>

```javascript
this.state = {
    ...
    errorMessage: "",
    isMaxLengthReached: false,
};

this.updateSearchValue = this.updateSearchValue.bind(this);
```

```javascript
updateSearchValue(value){
    if (value.length >= this.props.maxLength && !this.state.isMaxLengthReached){
        this.setState({
            isMaxLengthReached: true,
            errorMessage: this.props.translate("common.error.characterLimitExceed")
        });
        return;
    }

    if (this.state.isMaxLengthReached){
        if (value.length < this.props.maxLength && value.length > this.props.maxLength - 20){
            this.setState({
                errorMessage: this.props.translate("common.error.characterLimitExceedCounter", {length: value.length, limit: this.props.maxLength})
            });
        }else if (value.length <= this.props.maxLength - 20){
            this.setState({
                isMaxLengthReached: false,
                errorMessage: "",
            });
        }
    }

    this.props.onChangeText(value);
}
```
```javascript
<TextInput
    onChangeText={this.updateSearchValue}
    maxLength={this.props.maxLength}
    errorText={this.state.errorMessage}
                
 
 
 
<OptionsList
   ...
   headerMessage={Boolean(this.state.errorMessage) ? "" : this.props.headerMessage}           
```


</details>

But about `headerMessage` "no result fount"
we will hide it while the error without counter is displayed, because the max Length reached and no filtering process.
But while the counter is visible, and the max Length is not reached we still have filtering process, so I think it is not a good UI to hide it, and displaying both seem not the best UI.


https://github.com/Expensify/App/assets/41129870/481e773a-5d67-4a2f-a7b6-7cc5711f10c6


Any design idea about displaying `headerMessage` "no result fount" with the counter? @joekaufmanexpensify @allroundexperts ?

**Comment by ahmedGaber93** [comment]
Can adding the counter on the right side like composer?
![Screenshot 2023-09-21 at 12 12 20 AM](https://github.com/Expensify/App/assets/41129870/463a8cd4-5b00-4bce-89d8-bdddcc45b950)

**Comment by joekaufmanexpensify** [comment]
Hm, yeah that's interesting. @trjExpensify do you know if there is any precedent for showing two feedback/error messages at once? If so, then I like showing both the `no results found` and counter error simultaneously. 

But if not, then I think it's fine personally to not show the `no results found` while the user is over the first field limit.

**Comment by joekaufmanexpensify** [comment]
Cool cool, works for me. @allroundexperts any further input needed before we can proceed with this one?

**Comment by joekaufmanexpensify** [comment]
@allroundexperts bump on whether there's anything else you need here?

**Comment by johncschuster** [comment]
@allroundexperts bump!

**Comment by trjExpensify** [comment]
@allroundexperts @ahmedGaber93 addressing the video..

- why is there only a counter in the error message when under the limit and not over the limit as well? (Danny's screenshot for reference [here](https://github.com/Expensify/App/issues/25804#issuecomment-1705685194).
- I think the copy should be `Character limit exceeded (%count%/%maxCount%)`, right now it seems like it's using a plural instead of `character`

CC: @dannymcclain for feedback as well.

**Comment by dannymcclain** [comment]
Yeah I agree @tjferriss. The behavior in the video is a little confusing to me.

- Why is there an error message _without_ the counter at all? I think the point of the counter is to show that you are over the limit.
- If the number of characters is less than the limit (like how in the video it's 86/100), shouldn't the error go away?

I imagine it working like this (assuming a character limit of 100): 
- Enter 99 characters, no error
- Enter 100 characters, no error
- Enter 101 characters, error with text `Character limited exceeded (101/100)`
- Once character count goes back to 100 or less, the error goes away

I also think it makes perfect sense for the error message to replace the "No results found" message—in this situation you've entered an input that couldn't _possibly_ return any results because you've basically broken it haha. So it makes sense to me for the error message to replace the help text (I don't think anyone is arguing that it shouldn't, just wanted to toss my 2 cents in there)

**Comment by trjExpensify** [comment]
Poor @tjferriss, always getting tagged in GH. Agree with all that though 👍 

>Why is there an error message without the counter at all? 

Great point. Once the field is in error, we validate the input on change when editing it, so the error should go away once they are under the count. So in fact, in the above example with a keyboard smash it entered, it would return to "no results found" at that point.

**Comment by ahmedGaber93** [comment]
@trjExpensify 
> why is there only a counter in the error message when under the limit

I add the proposal based on this comment https://github.com/Expensify/App/issues/25804#issuecomment-1726788809 as I think this is the last of discussion.

But no problem, I will add a new proposal base on @dannymcclain https://github.com/Expensify/App/issues/25804#issuecomment-1743643174

Thanks!

**Comment by ahmedGaber93** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/25804#issuecomment-1690722686)


https://github.com/Expensify/App/assets/41129870/6f4ac516-8188-4f54-945b-b1e02a975f41

**Comment by dannymcclain** [comment]
This is looking good to me too! Can we just make sure there's a space between the text and the counter please. In the video it's looking like `Character limit exceeded(520/500)` and we just want it to be `Character limit exceeded (520/500)`

**Comment by ahmedGaber93** [comment]
@allroundexperts PR https://github.com/Expensify/App/pull/28722 is ready for review.

**Comment by joekaufmanexpensify** [comment]
Now that we've moved forward here, I feel like we are in a good place to begin using this error pattern in other related issues that were waiting on this discussion (like https://github.com/Expensify/App/issues/18647 and https://github.com/Expensify/App/issues/19821). Unless anyone disagrees, going to update those issues so we can move forward on them!

**Comment by trjExpensify** [comment]
Makes sense! Though I question whether we need a PR for #19821. I think implementing #18647 incl. the character counter from this issue will take care of it.

**Comment by joekaufmanexpensify** [comment]
Ah, good call. The character counter we are implementing here will apply to every field in the app that has an existing character limit by default?

**Comment by trjExpensify** [comment]
I'm not sure, tbh. But doesn't #18647 introduce the character limits for fields, and therefore if the character counter doesn't get included by default, the PR for that issue should incorporate it.

**Comment by joekaufmanexpensify** [comment]
In #18647 , we decided on the standard field length limits, and then are [updating the existing limits](https://expensify.slack.com/archives/C01GTK53T8Q/p1690209885486499?thread_ts=1686863510.616919&cid=C01GTK53T8Q) for a few fields to match them. I agree we would add the character counter for those fields. 

We had [actually decided](https://expensify.slack.com/archives/C01GTK53T8Q/p1691747854865349?thread_ts=1686863510.616919&cid=C01GTK53T8Q) to leave the limit for legal name as is during the original discussion though. So we had not been planning to make any changes to legal name in #18647.

**Comment by trjExpensify** [comment]
Ah okay cool, it's staying as is. Cool, so then we just need to implement the character counter on: https://github.com/Expensify/App/issues/19821 then.

**Comment by joekaufmanexpensify** [comment]
Yep, makes sense to me!

**Comment by flodnv** [comment]
Maybe? What do you think @johncschuster ?

**Comment by thuyle04** [comment]
Contributor details
Your Expensify account email: [thuy.lethithu@antbuddy.com](mailto:thuy.lethithu@antbuddy.com)
Upwork Profile Link: https://www.upwork.com/freelancers/~01fcd9ed589276332f

**Comment by thuyle04** [comment]
Contributor details
Your Expensify account email: thuy.lethithu@antbuddy.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01fcd9ed589276332f

**Comment by JmillsExpensify** [comment]
Need a payment summary for this issue.

**Comment by johncschuster** [comment]
## Payment Summary
@allroundexperts does not require payment (Eligible for Manual Requests) - $1000 - paid via NewDot
@ahmedGaber93 requires payment -$1000 - paid via Upwork
@LeThiThuThuy requires payment - $250 (previous Bug Reporter payment amount) - paid via Upwork

**Comment by JmillsExpensify** [comment]
$1,000 payment approved for @allroundexperts based on summary above.

**Comment by johncschuster** [comment]
@allroundexperts for the regression test, I think this is a typo:

> Paste text that exceeds 500k characters into the search bar

That's just meant to be "exceeds 500 characters", right?

**Comment by thuyle04** [comment]
Contributor details
Your Expensify account email: thuy.lethithu@antbuddy.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01fcd9ed589276332f

**Comment by johncschuster** [comment]
Looks like the job closed on Upwork. I'll create a payment issue and add @ahmedGaber93 and @LeThiThuThuy to that.

**Comment by johncschuster** [comment]
Ah, it looks like @LeThiThuThuy has been updated to @thuyle04. I will add them to the payment issue.

**Comment by johncschuster** [comment]
@ahmedGaber93, please comment on the above GH so I can assign it to you for payment. Thank you!

**Comment by ahmedGaber93** [comment]
Ok, commented.

**Comment by thuyle04** [comment]
Contributor details
Your Expensify account email: thuy.lethithu@antbuddy.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01fcd9ed589276332f

**Comment by johncschuster** [comment]
@ahmedGaber93 I don't see your comment on https://github.com/Expensify/App/issues/30399. Can you please try again? Once you've commented, I can assign it and you'll be invited to the job.

**Comment by ahmedGaber93** [comment]
Ok, commented recheck @johncschuster

**Comment by johncschuster** [comment]
Payment is being issued [here](https://github.com/Expensify/App/issues/30399). I'm going to close this issue.
